Add exp validation and respond with a 401 when a token is expired

quantumew opened the pull request:
Related https://github.com/articulate/authentic-rb/pull/7

We had another task to ensure we were validating the exp claim so this adds that here as well. The JWT library I chose doesn't do any claim value validation, just ensures they are signed.

Commit:
Add exp validation and respond with a 403 when a token is expired

dbertram reviewed (approved):
This is the first `go` code I've ever reviewed, but it lgtm.

I'd suggest getting someone who knows `go` to review for style/convention though :)

chazu reviewed (approved):
Looks good to me - tests are green. :shipit:

Commit:
Updating respons code for expired token to 401

I think being able to distinguish is still beneficial but it turns out
the RFC does require a 401 for invalid vs expired.

dbertram commented on main_test.go:
```diff
@@ -178,14 +211,20 @@ var _ = Describe("authentic", func() {
 				Expect(body.Message).To(Equal("Unauthorized"))
 			})
 
+			It("returns 403 response in Gin middleware due to expired token", func() {
```
403 => 401 in description?

quantumew replied:
Ah yea good catch e88b57937c832a793e0ee5307b9801af9d0cae90

Commit:
Fix context text

quantumew commented:
To clarify I misunderstood the RFC second hand. 401 on invalid/expired token (although distinguishing the two will provide some value as far as debugging goes), and a 403 when the token is good but the user does not have the necessary access level, scope/aud issue.
